File: LeetCode/5회차/B20-20-Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        table = {
            '(': ')',
            '{': '}',
            '[': ']'
        }

        stack = []
        for char in s:
            if char in table:
                stack.append(char)
            else:
                if not stack or char != table[stack.pop()]:
                    return False

        return len(stack) == 0


print(Solution().isValid("()[]{}"))

File: LeetCode/JavaScript/B20-20-Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const table = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  stack = new Array();

  for (c of s) {
    if (!!table[c]) {
      stack.push(c);
    } else {
      if (stack.length === 0 || c !== table[stack.pop()]) {
        return false;
      }
    }
  }

  return !stack.length;
};
